sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -1344,7 +1360,13 @@ case class Sample(
       s"Sampling fraction ($fraction) must be on interval [0, 1] without replacement")
   }
 
-  override def maxRows: Option[Long] = child.maxRows
```

Comment:
I am not sure whether it is wrong, sampling with replacement should not generate more rows than the input dataset.

But we can not impl a strict sampling with replacement, so `PoissonSampler` is used instead, which can not guarantee this attribute.

```
scala> val df = spark.range(0, 1000)
df: org.apache.spark.sql.Dataset[Long] = [id: bigint]

scala> df.count
res0: Long = 1000

scala> df.sample(true, 0.999999, 10).count
res1: Long = 1004
```

Reply:
interesting, cc @sigmod @maryannxue @srielau 

this seems like a correctness issue and we should fix it ASAP. @zhengruifeng can you open a new PR for the sample fix only so that we can backport?

Reply:
@cloud-fan   Sure